### PR TITLE
Virtual work

### PR DESCRIPTION
**Pull request by sslattery**
These changes add a transpose apply option for the operators and enables it for operators where possible. The transpose of the moving least square operator is verified to satisfy conservation of virtual work in an added test.

### CODE REVIEW

File: packages/Interface/src/OperatorVector/DTK_MapOperator.hpp [hunk 1]
```diff
@@ -123,11 +123,14 @@ class MapOperator : public Tpetra::Operator<double,int,SupportId>
...
 	double alpha = Teuchos::ScalarTraits<double>::one(),
 	double beta = Teuchos::ScalarTraits<double>::zero() ) const override;
-    bool hasTransposeApply() const override { return false; }
+    bool hasTransposeApply() const override;
     //@}
 
   protected:
 
+    //! Tranpose apply option.
+    virtual bool hasTransposeApplyImpl() const = 0;
```
Comment (dalg24) [inline packages/Interface/src/OperatorVector/DTK_MapOperator.hpp]: Why didn't you just declare `MapOperator::hasTransposeApply()` pure virtual?

Comment (dalg24) [inline packages/Interface/src/OperatorVector/DTK_MapOperator.hpp]: Yes that's what I was looking at. It made more sense for ``apply(...)`` and ``setup(...)`` but consistency is a fine reason. No worries.

File: packages/Operators/test/tstVirtualWork.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    // Get the communicator.
+    Teuchos::RCP<const Teuchos::Comm<int> > comm =
+	Teuchos::DefaultComm<int>::getComm();
+    int comm_rank = comm->getRank();
+    int comm_size = comm->getSize();
+    int inverse_rank = comm_size - comm_rank - 1;
+    const int space_dim = 3;
+
+    // Seed the RNG
+    std::srand( 324903231 );
```
Comment (dalg24) [inline packages/Operators/test/tstVirtualWork.cpp]: Couldn't help myself to do
```
#include <iostream>
#include <limits>
int main()
{
  std::cout<< std::numeric_limits<unsigned int>::max() << std::endl;
  return 0;
}
```
which prints `4294967295`. We are good :D

File: packages/Operators/test/tstVirtualWork.cpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        Teuchos::rcp( new DataTransferKit::FieldMultiVector(
+                          structure_forces_field,
+                          structure_manager.functionSpace()->entitySet()) );
+
+    // Create the point cloud operator to map from the structure to the fluid.
+    DataTransferKit::MapOperatorFactory factory;
+    Teuchos::RCP<DataTransferKit::MapOperator> cloud_op =
+        factory.create( structure_displacements_vector->getMap(), 
+                        fluid_displacements_vector->getMap(), 
+                        *parameters );
```
Comment (dalg24) [inline packages/Operators/test/tstVirtualWork.cpp]: I had forgotten that this silly ``MapOperatorFactory`` doesn't have create as a static member function. Sigh...

File: packages/Operators/test/tstVirtualWork.cpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    double structure_virtual_work = 0.0;
+    double fluid_virtual_work = 0.0;
+    setupAndRunTest(
+	"mls_test_knn.xml", structure_virtual_work, fluid_virtual_work );
+    
+    // Check the results.
+    TEST_FLOATING_EQUALITY( fluid_virtual_work, 
+                            structure_virtual_work, 
+                            epsilon );
+}
```
Comment (dalg24) [inline packages/Operators/test/tstVirtualWork.cpp]: That's where I think Boost.Test is superior to the Teuchos Unit Testing harness with the [data-driven test cases](http://www.boost.org/doc/libs/1_61_0/libs/test/doc/html/boost_test/tests_organization/test_cases/test_case_generation.html).

**Comment by dalg24** [comment]
Let's finish up work on PR #70 and rebase. This PR is ready.

**Comment by sslattery** [comment]
so do you want to merge this one then? I ended up branching from master instead of the other branch

**Comment by dalg24** [comment]
Sure.

**Pull request merged**